provider/storage_benchmark.py

```diff
@@ -337,20 +337,25 @@ def run(self, cmd_options='-a', timeout=1800):
 
 class FioLinuxCfg(object):
     def __init__(self, params, session):
-        fio_pkg = params.get("fio_pkg", 'fio-3.13-48-ga819.tar.bz2')
-        host_path = os.path.join(data_dir.get_deps_dir(), 'fio', fio_pkg)
-        self.download_path = os.path.join('/home', fio_pkg)
-        self.fio_inst = os.path.join('/home', 'fio_inst')
-        self.fio_path = '%s/bin/fio' % self.fio_inst
-        scp_benckmark = attrgetter('scp_benckmark')
-        unpack_file = attrgetter('unpack_file')
-        install_timeout = params.get_numeric('fio_install_timeout', 300)
-        install = attrgetter('install')
-        self.setups = {scp_benckmark: (params.get('username'), params.get('password'),
-                                       host_path, self.download_path),
-                       unpack_file: (TAR_UNPACK, self.download_path, self.fio_inst),
-                       install: (self.fio_inst, self.fio_inst, install_timeout)}
-        self.setup_orders = (scp_benckmark, unpack_file, install)
+        status, output = session.cmd_status_output('which fio')
+        if not status:
+            self.fio_path = output.strip()
+            self.setup_orders = ()
+        else:
+            fio_pkg = params.get("fio_pkg", 'fio-3.13-48-ga819.tar.bz2')
+            host_path = os.path.join(data_dir.get_deps_dir(), 'fio', fio_pkg)
+            self.download_path = os.path.join('/home', fio_pkg)
+            self.fio_inst = os.path.join('/home', 'fio_inst')
+            self.fio_path = '%s/bin/fio' % self.fio_inst
+            scp_benckmark = attrgetter('scp_benckmark')
+            unpack_file = attrgetter('unpack_file')
+            install_timeout = params.get_numeric('fio_install_timeout', 300)
+            install = attrgetter('install')
+            self.setups = {scp_benckmark: (params.get('username'), params.get('password'),
+                                           host_path, self.download_path),
+                           unpack_file: (TAR_UNPACK, self.download_path, self.fio_inst),
+                           install: (self.fio_inst, self.fio_inst, install_timeout)}
+            self.setup_orders = (scp_benckmark, unpack_file, install)
```

Comment:
From this solution, I see the priority of default value is higher then custom value, this it make value which caller would like meaningless. And only from my side, I think the requirements from callers should be higher then default since APIs are server for callers.
There is another scenario I consider that if the fio existed has some issue, the latest fio is ok at this time, how could we resolve it? 
And if the upstream caller want to test their cases by they compiled fio tool that added some new feature test functions in their fio tool but they do not want this tool as public, but now just only use the default fio tool. so for this situation, maybe not work.

```suggestion
        fio_pkg = params.get("fio_pkg")
        if fio_pkg:
            host_path = os.path.join(data_dir.get_deps_dir(), 'fio', fio_pkg)
            self.download_path = os.path.join('/home', fio_pkg)
            self.fio_inst = os.path.join('/home', 'fio_inst')
            self.fio_path = '%s/bin/fio' % self.fio_inst
            scp_benckmark = attrgetter('scp_benckmark')
            unpack_file = attrgetter('unpack_file')
            install_timeout = params.get_numeric('fio_install_timeout', 300)
            install = attrgetter('install')
            self.setups = {scp_benckmark: (params.get('username'), params.get('password'),
                                           host_path, self.download_path),
                           unpack_file: (TAR_UNPACK, self.download_path, self.fio_inst),
                           install: (self.fio_inst, self.fio_inst, install_timeout)}
            self.setup_orders = (scp_benckmark, unpack_file, install)
        else:
            status, output = session.cmd_status_output('which fio')
            if not status:
                self.fio_path = output.strip()
                self.setup_orders = ()            
```

Reply:
> From this solution, I see the priority of default value is higher then custom value, this it make value which caller would like meaningless. And only from my side, I think the requirements from callers should be higher then default since APIs are server for callers.
> There is another scenario I consider that if the fio existed has some issue, the latest fio is ok at this time, how could we resolve it?
> And if the upstream caller want to test their cases by they compiled fio tool that added some new feature test functions in their fio tool but they do not want this tool as public, but now just only use the default fio tool. so for this situation, maybe not work.

I can not say which fio resource has high priority is better, it depends ~   
From my view, the value in the distro is more stable than the manual compiled binary,  but i agree with you that we need consider the situation that caller want to use the specified fio tools as different test requirement.

I update the patch to make the priority configurable, please help re-review, thanks a lot!